### PR TITLE
[CodeCompletion] Modifier related improvements in CompletionOverrideLookup

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?
- If `required` or `convenience` is specified, emit only initializers
- If `final` or `open` is specified, don't emit initializers or typealias
- If `typealias` is specified, emit only associated type implementation
- Emit `override` or `required` modifier for initializers
- Emit access modifier for initializers
- Emit designated initializers even if `override` is specified
- Don't emit inheritance clause for associated type implementation ([SR-2383](https://bugs.swift.org/browse/SR-2383))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please clean test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please clean test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |
| Linux platform | @swift-ci Please clean test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@benlangmuir Could you review this?

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -4243,30 +4282,37 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
...
         addMethodOverride(FD, Reason);
       return;
     }
 
     if (auto *VD = dyn_cast<VarDecl>(D)) {
-      if (!hasIntroducer || hasVarIntroducer) {
+      if (hasInitializerModifier)
+        return;
+
+      if (!hasIntroducer || hasVarIntroducer)
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: If someone wrote `required var` or `convenience func`, etc. I suggest we still give them var and func override completions.  It's going to be invalid, but they can accept the fixits.  So something like
```c++
if (hasVarIntroducer || (!hasIntroducer && !hasInitializerModifier)) {
```
and likewise for functions.

**Comment by benlangmuir** [comment]
Otherwise, LGTM!

File: lib/IDE/CodeCompletion.cpp (hunk 1 above)
Comment (rintaro) [inline lib/IDE/CodeCompletion.cpp]: 👍  Makes sense!
And also for typealias. e.g. `final typealias`

**Comment by benlangmuir** [comment]
@swift-ci Please smoke test

**Comment by benlangmuir** [comment]
@swift-ci Please smoke test linux

**Pull request merged**